src/MiniJuvix/Translation/MonoJuvixToMiniC.hs

```diff
@@ -18,57 +18,127 @@ newtype MiniCResult = MiniCResult
   { _resultCCode :: Text
   }
 
+data CFunType = CFunType
+  { _cFunArgTypes :: [CDeclType],
+    _cFunReturnType :: CDeclType
+  }
+  deriving stock (Show, Eq)
+
+data BindingInfo = BindingInfo
+  { _bindingInfoExpr :: Expression,
+    _bindingInfoType :: CFunType
+  }
+
+newtype PatternInfoTable = PatternInfoTable
+  {_patternBindings :: HashMap Text BindingInfo}
+
+data ClosureInfo = ClosureInfo
+  { _closureNameId :: Mono.NameId,
+    _closureRootName :: Text,
+    _closureFunType :: CFunType
+  }
+  deriving stock (Show, Eq)
+
 makeLenses ''MiniCResult
+makeLenses ''CFunType
+makeLenses ''PatternInfoTable
+makeLenses ''BindingInfo
+makeLenses ''ClosureInfo
 
-entryMiniC ::
-  Mono.MonoJuvixResult ->
-  Sem r MiniCResult
+entryMiniC :: Mono.MonoJuvixResult -> Sem r MiniCResult
 entryMiniC i = return (MiniCResult (serialize cunitResult))
   where
     compileInfo :: Mono.CompileInfoTable
     compileInfo = Mono.compileInfoTable i
+
     cunitResult :: CCodeUnit
     cunitResult =
       CCodeUnit
         { _ccodeCode = cheader <> cmodules
         }
+
     cheader :: [CCode]
     cheader =
       map
         ExternalMacro
         [ CppIncludeSystem Str.stdbool,
           CppIncludeFile Str.minicRuntime
         ]
+
     cmodules :: [CCode]
-    cmodules = toList (i ^. Mono.resultModules) >>= (run . runReader compileInfo . goModule)
+    cmodules = do
+      m <- toList (i ^. Mono.resultModules)
+      let buildTable = Mono.buildTable m
+      run (runReader compileInfo (genCTypes m))
+        <> genFunctionSigs m
+        <> run (runReader buildTable (genClosures m))
+        <> run (runReader buildTable (genFunctionDefs m))
 
 unsupported :: Text -> a
 unsupported msg = error (msg <> " Mono to C: not yet supported")
 
-goModule ::
-  Member (Reader Mono.CompileInfoTable) r =>
+genCTypes :: forall r. Members '[Reader Mono.CompileInfoTable] r => Mono.Module -> Sem r [CCode]
+genCTypes Mono.Module {..} =
+  concatMapM go (_moduleBody ^. Mono.moduleStatements)
+  where
+    go :: Mono.Statement -> Sem r [CCode]
+    go = \case
+      Mono.StatementInductive d -> return (goInductiveDef d)
+      Mono.StatementAxiom d -> goAxiom d
+      Mono.StatementForeign d -> return (goForeign d)
+      Mono.StatementFunction {} -> return []
+
+genFunctionSigs :: Mono.Module -> [CCode]
+genFunctionSigs Mono.Module {..} =
+  go =<< _moduleBody ^. Mono.moduleStatements
+  where
+    go :: Mono.Statement -> [CCode]
```

Comment:
I counted, here and below, at least three functions using the same pattern, pattern-matching on M.Statement and applying another function only on the function statement case. So, is it worth to consider using a function like the following? 

```
applyOnFunStatement :: forall a.
  Monoid a =>  (Mono.FunctionDef -> a) -> Mono.Statement -> a
applyOnFunStatement f = \case
  Mono.StatementFunction x -> (f x)
  Mono.StatementForeign {} -> mempty
  Mono.StatementAxiom {} -> mempty
  Mono.StatementInductive {} -> mempty
``` 
Then, we will define `go` as `applyOnFunStatement genFunctionSig`.

Reply:
Yes this is a good idea - but this code has changed substantially in the work for closure environments I've been doing on-top of this PR. So I'd prefer not to change this code in this PR to avoid conflicts.

Reply:
I agree.